### PR TITLE
Vi/an/user account simple

### LINKED ISSUE #3: USER ACCOUNT - Simple login - Backend
The user can login using a username and password.

NOTE: 
- this functionality is primarily for testing purposes and will be deleted at the end of the project.

### LINKED ISSUE #5: USER ACCOUNT - User name
The user should be able to choose a unique name that will be displayed on the website.

### PR DESCRIPTION
**Pull request by vincentvis**
Quite big...

- Adds "SECRET" to .env for `jwt` config (SECRET is a key that is used in hashing tokens)
- Defines the user entity
  - we could consider putting restrictions on username (max-length, ascii only etc?)
  - we could consider putting requirements on the password (min-length etc) (good for security, annoying for testing..)
- Adds User controller responsible for user related things
- Adds Auth controller responsible for login/logout
- Bunch of tests for both User and Auth (both unit and e2e)

closes #3 
closes #5

### CODE REVIEW

File: backend/src/user/user.controller.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import { AuthGuard } from '../auth/auth.guard';
+import { UpdateUserDto } from './dto/update-user.dto';
+import { InsertResult, QueryFailedError, UpdateResult } from 'typeorm';
+import { CreateUserDto } from './dto/create-user.dto';
+import { RegisterUserDto } from './dto/register-user.dto';
+
+@UseInterceptors(ClassSerializerInterceptor)
+@Controller('users')
+export class UserController {
+	constructor(
```
Comment (Ryno95) [inline backend/src/user/user.controller.ts:26]: Is the `// authService` comment still needed?

File: backend/src/user/user.controller.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+				updateUserDto,
+			);
+			return updateResult;
+		} catch (e) {
+			if (e instanceof QueryFailedError) {
+				throw new BadRequestException('Please pick a different username');
+			}
+			if (e instanceof NotFoundException) {
+				throw e;
+			}
```
Comment (Ryno95) [inline backend/src/user/user.controller.ts:80]: console.log on line 81, specifically to check what the error was if not one of the above? or was it fo dev purposes?

File: backend/test/auth.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+import { UserModule } from '../src/user/user.module';
+import { SharedModule } from '../src/shared/shared.module';
+import { AuthModule } from '../src/auth/auth.module';
+import { UserService } from '../src/user/user.service';
+import { AuthService } from '../src/auth/auth.service';
+import { User } from '../src/user/entities/user.entity';
+import { InsertResult } from 'typeorm';
+import { JwtService } from '@nestjs/jwt';
+import * as cookieParser from 'cookie-parser';
+
```
Comment (Ryno95) [inline backend/test/auth.e2e-spec.ts:24]: Animal Controller in describe description ;)

File: backend/test/auth.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
+import { Test, TestingModule } from '@nestjs/testing';
```
Comment (Ryno95) [inline backend/test/auth.e2e-spec.ts:1]: Unused imports and or vars:

- Body
- AuthController and AuthService

**Review by Ryno95 — COMMENTED** [review]
Solid edge case testing for the user! Only some comments/naming/console change requests :)

File: backend/src/user/user.controller.ts (hunk 1 above)
Comment (vincentvis) [reply backend/src/user/user.controller.ts:26]: Definitely not!

File: backend/src/user/user.controller.ts (hunk 2 above)
Comment (vincentvis) [reply backend/src/user/user.controller.ts:78]: I honestly have no clue why it would be any other than one of these two errors (QueryFailed/NotFound)
So It's there for us to see if ever anything else goes wrong.. I'll add a piece of text to make that more clear

**Review by Ryno95 — APPROVED** [review]

File: backend/test/dotenv-config.ts [hunk 5]
```diff
@@ -1,3 +1,3 @@
 require('dotenv').config({
-  path: '.env.test',
+	path: '.env.test',
```
Comment (livlaming) [inline backend/test/dotenv-config.ts:2]: not sure what happens here? (surely it is corretct, but I'm curious!)

File: backend/src/auth/auth.controller.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+		}
+	}
+
+	@UseGuards(AuthGuard)
+	@Post('logout')
+	@HttpCode(HttpStatus.OK)
+	async logout(@Res({ passthrough: true }) response: Response) {
+		response.clearCookie('jwt');
+		return {
+			message: 'Success',
```
[the remark this answers is not in the document]
Comment (vincentvis) [reply backend/src/auth/auth.controller.ts:63]: Good one! Technically, I guess, yes, it could fail...
source of clearCookie form expressjs: https://github.com/expressjs/express/blob/158a17031a2668269aedb31ea07b58d6b700272b/lib/response.js#L824:L887

It has a `throw`, but only for signed cookies.. 
And it uses JSON.stringify, which could also throw.. but unlikely in this use case [(throws when you give it a object with circular reference or when out of memory..)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify)

The case against protecting against this is that expressjs in their own tests, do not check for errors:
https://github.com/expressjs/express/blob/a65913776d0b16837364ee66caa1a7f38a9997c0/test/res.clearCookie.js

But that could be a lack of testing from their side..... who knows 🤷

File: backend/src/auth/auth.controller.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		};
+	}
+
+	@UseGuards(AuthGuard)
+	@Get('me')
```
[the remark this answers is not in the document]
Comment (vincentvis) [reply backend/src/auth/auth.controller.ts:68]: I wasn't sure what to call this.. But this gets user data about the currently logged in user, which i've seen in other API's where they use the /me endpoint..

File: backend/src/auth/auth.controller.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import { UserService } from '../user/user.service';
+import * as bcrypt from 'bcrypt';
+import { JwtService } from '@nestjs/jwt';
+import { Response, Request } from 'express';
+import { AuthGuard } from './auth.guard';
+import { AuthService } from './auth.service';
+import { LoginUserDto } from './dto/login-user.dto';
+
+@UseInterceptors(ClassSerializerInterceptor)
+@Controller()
```
[the remark this answers is not in the document]
Comment (vincentvis) [reply backend/src/auth/auth.controller.ts:24]: this registers these routes:
`/api/login/`
`/api/logout/`
`/api/me/`

Which are quite clear on their own and I don't think wrapping them with another /auth/ would make it much clearer, so here I'd opt for shorter :)

File: backend/src/shared/shared.module.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import { Module } from '@nestjs/common';
+import { ConfigModule, ConfigService } from '@nestjs/config';
+import { JwtModule } from '@nestjs/jwt';
+
```
[the remark this answers is not in the document]
Comment (vincentvis) [reply backend/src/shared/shared.module.ts:4]: I think the shared module is meant to handle imports etc that are shared between multiple other modules but require some initialization
So the JWT module requires some setup, but that setup should only be ran once. This makes sure it's only ran once but can still be used in multiple parts of the project.

[the remark this answers is not in the document]
Comment (vincentvis) [reply backend/src/shared/shared.module.ts:4]: I'm not sure what you mean with the decorator, since there is no special decorator in this file?

File: backend/src/user/entities/user.entity.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+	UpdateDateColumn,
+} from 'typeorm';
+import { Exclude } from 'class-transformer';
+
+@Entity('users')
+export class User {
+	@PrimaryGeneratedColumn()
+	id: number;
+
+	@Column({ default: true })
```
[the remark this answers is not in the document]
Comment (vincentvis) [reply backend/src/user/entities/user.entity.ts:15]: I think this default is fine.
The subject wants that we do all users through Intra. 
The fact that we do "simple" users is our own addition. So having intra as default made sense to me. 
If it causes problems in the register/login flow for simple_users vs intra_users we can change it then :)

File: backend/src/user/entities/user.entity.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+	is_intra: boolean;
+
+	@Column({ unique: true, nullable: false })
```
[the remark this answers is not in the document]
Comment (vincentvis) [reply backend/src/user/entities/user.entity.ts:18]: Couldn't really find a good way to do this, but there was a weird thing that you could save a user without a username... (not an empty string, just a missing paramter)
this solved that :P

File: backend/test/dotenv-config.ts (hunk 5 above)
Comment (vincentvis) [reply backend/test/dotenv-config.ts:2]: this is pretterjs + some config changing the spaces to a tab

File: backend/src/auth/auth.controller.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (livlaming) [reply backend/src/auth/auth.controller.ts:24]: agreed!

File: backend/src/user/entities/user.entity.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (livlaming) [reply backend/src/user/entities/user.entity.ts:15]: cool

File: backend/src/user/entities/user.entity.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (livlaming) [reply backend/src/user/entities/user.entity.ts:18]: whatever works right :)

File: backend/test/dotenv-config.ts (hunk 5 above)
Comment (livlaming) [reply backend/test/dotenv-config.ts:2]: thanks

File: backend/src/auth/auth.controller.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (livlaming) [reply backend/src/auth/auth.controller.ts:75]: fine by me hehe
